### DISCUSSION TITLE
Investigate viability of larger time steps for JuPedSim

### DISCUSSION THREAD

**Discussion by Carole1004** [PedestrianDynamics/jupedsim#1306] [General]
**Bottleneck**
We start the study with the well-known case of a bottleneck and the default parameters of the speed model. We observe via the JuPedSim visualizer that agents are being pushed out of the walkable area, i.e. that there are intersections between agents and the outside, which should be prevented by choosing  the correct model parameters. In order to quantify the problem, we count and locate the intersections. The intersections are located at the border of the spawning area as well as at the wall near the bottleneck itself. Those are the places where the geometry is the most crowded. The exact number of intersections is around 2.000 even for reference time step dt = 0.01 s.
An option to deal with this problem is to vary those parameters from the Collision Free Speed Model:

1. Strength of the repulsion from geometry boundaries: strength_geometry_repulsion = 5.0 (default parameter)
2. Range of the repulsion from geometry boundaries: range_geometry_repulsion = 0.02 (default parameter)

We vary those parameters and check, if the agents are staying inside of the walkable area with the reference time step  dt = 0.01 s. We obtain good results with the following values:

1. strength_geometry_repulsion = 26
2. range_geometry_repulsion = 0.01
--> Those are the new reference parameters of our study.

In order to increase the time step successfully, the intersections of the agents with the outside of the walkable area needs to be taken in to account. Several parameter sets were investigated and the best result with dt = 0.02 s is with:

1. strength_geometry_repulsion = 35
2. range_geometry_repulsion = 0.019

_Simulation was run with this notebook:_ 
[Schrittweite_Bneck_35_0-019_dt_0-02.ipynb.zip](https://github.com/PedestrianDynamics/jupedsim/files/13858963/Schrittweite_Bneck_35_0-019_dt_0-02.ipynb.zip)

**Comment by Carole1004** [comment]
In the following we will compare the trajectories, the N-t grafics and the Voronoi densities for both set of parameters.

dt = 0.01 s is on the left and dt = 0.02 s is on the right:

<img width="977" alt="Screenshot 2023-12-20 at 11 28 56" src="https://github.com/PedestrianDynamics/jupedsim/assets/144786992/4dd1ca01-1587-4d2e-a12e-2f74845e376a">

Even if the trajectories and the N-t grafics are not exactly the same, no distinct differences are observed. Anyway, the complete duration of simulation is a few seconds longer for dt = 0.02 s.

**Comment by Carole1004** [comment]
<img width="1008" alt="Screenshot 2023-12-20 at 11 31 23" src="https://github.com/PedestrianDynamics/jupedsim/assets/144786992/d27460be-a67a-4d89-b824-e2df767c0d2d">

The measurement area (blue square) is located in front of the bottleneck, where the density is the highest. The profile of the Voronoi density is not exactly the same, but no disitinct difference is observed.

**Comment by Carole1004** [comment]
The next results are the density and speed profiles of both sets of parameters between t = 20 s and t = 40 s. No distinct difference is observed here between the reference (dt = 0.01 s) and the new studied parameters (dt = 0.02 s).

<img width="904" alt="Screenshot_2023-12-20_at_11 32 59" src="https://github.com/PedestrianDynamics/jupedsim/assets/144786992/6096042c-f383-42b8-9061-c36195427d44">

**Comment by Carole1004** [comment]
Furthermore, the simulation has been carried out for different seeds. Again, no distinct difference is observed - only the number of agents appearing in the spawning area.

<img width="1149" alt="Screenshot 2023-12-20 at 11 35 51" src="https://github.com/PedestrianDynamics/jupedsim/assets/144786992/ea6b4648-cd22-4bb1-b337-6095d0a0a980">

**Comment by Carole1004** [comment]
Flow for different seeds:

<img width="1057" alt="Screenshot 2023-12-20 at 11 37 15" src="https://github.com/PedestrianDynamics/jupedsim/assets/144786992/f3c78af4-ae11-4832-8bcd-6c293e40a008">

**Comment by Carole1004** [comment]
Comparison with results from [Seyfried 2009]:  The flow through the bottleneck measured with our simulations (see tables above)  are between 2.42/s and 2.57/s for a time step of 0.01 s  and they are between 2.31/s and 2.38/s for a time step of 0.02 s.  Our results are in accordance with the results shown in the paper for b = 1.2 m.

<img width="380" alt="Screenshot_2023-12-20_at_11 42 44" src="https://github.com/PedestrianDynamics/jupedsim/assets/144786992/29b5d20e-b542-44eb-942e-b30764910f89">

**Comment by Carole1004** [comment]
**Corner**
The second geometry in our study is the corner. Here again, we compare the results of the reference (dt = 0.01 s on the left) with the parameters for the study (dt = 0.02 s on the right). The profile of the trajectories just before the exit is a bit different due to two or three agents, which is negligible.

_Simulation was run with this notebook:_ 
[Schrittweite_Corner_35_0-019_dt_0-02.ipynb.zip](https://github.com/PedestrianDynamics/jupedsim/files/13859586/Schrittweite_Corner_35_0-019_dt_0-02.ipynb.zip)

<img width="862" alt="Screenshot 2023-12-20 at 11 44 44" src="https://github.com/PedestrianDynamics/jupedsim/assets/144786992/21f1eb6f-eda0-42bf-8021-0b9c2ff2b36b">

**Comment by Carole1004** [comment]
Next, we observe the density and speed profiles of both sets of parameters from the beginning to the end of the simulation. No distinct difference is observed here between time steps of dt = 0.01 s  and dt = 0.02 s.

<img width="707" alt="Screenshot 2023-12-20 at 11 46 34" src="https://github.com/PedestrianDynamics/jupedsim/assets/144786992/59019090-c690-4776-b4ec-36fa513ec7e7">

**Comment by Carole1004** [comment]
Again, the simulation has been carried out with different seeds. Apart from the number of agents in the spawning area, no distinct difference is observed.


<img width="993" alt="Screenshot 2023-12-20 at 11 47 10" src="https://github.com/PedestrianDynamics/jupedsim/assets/144786992/c39085b4-1dad-4a2c-a3cc-7ee168083629">

**Comment by Carole1004** [comment]
**Two-way-corridor**
The third geometry is a two-way-corridor with 25 agents coming from the left and 25 agents coming from the right. Here again, the results of our reference (the profiles with the time step dt = 0.01 s are up) are compared with the parameters of the study (the profiles with the time step dt = 0.02 s are below) between the times t = 10 s and t = 20 s. No distinct difference is observed.

Simulation was run with this notebook: 
[Schrittweite_2-way-corridor_35_0-019_dt_0-02.ipynb.zip](https://github.com/PedestrianDynamics/jupedsim/files/13859710/Schrittweite_2-way-corridor_35_0-019_dt_0-02.ipynb.zip)

<img width="1067" alt="Screenshot 2023-12-20 at 11 48 09" src="https://github.com/PedestrianDynamics/jupedsim/assets/144786992/f19e0599-e532-4520-a422-1bef7c60c608">

**Comment by Carole1004** [comment]
Let’s see if we can gain more information from the positions of the agents at different times: t = 0 s, t = 17 s and t = 23 s, for the time step dt = 0.01 s (up) and the time step dt = 0.02 s (below). 
At t = 17 s: lanes are formed for both scenarios.

<img width="1039" alt="Screenshot 2023-12-20 at 11 49 35" src="https://github.com/PedestrianDynamics/jupedsim/assets/144786992/7ee4a03c-c3f1-40f4-a8c4-38b599038024">

**Comment by Carole1004** [comment]
Simulations with different seeds: no distinct differences. 

<img width="960" alt="Screenshot 2023-12-20 at 11 50 22" src="https://github.com/PedestrianDynamics/jupedsim/assets/144786992/a92f1d77-adf5-490d-95d1-cd1a9f341ce7">

**Comment by Carole1004** [comment]
**Conclusion**: those new parameters seem to be suitable for our purpose.

**Comment by Ozaq** [comment]
@JetteSchumann / @Carole1004 / @chraibi  quick clarification:

To execute with a step size of 0.02s you suggest to use:
```
strength_neighbour_repulsion = default
range_neightbour_repulsion = default
strength_geometry_repulsion = 35
range_geometry_repulsion = 0.019
```

For the reference you used
```
strength_neighbour_repulsion = default
range_neightbour_repulsion = default
strength_geometry_repulsion = 26
range_geometry_repulsion = 0.01
```

Why did you modify the defaults for the comparison? Id assume because there have been issues with geometry collisions, right? In that case this indicates to me that we should update the defaults for geometry repulsion.

If the set for 0.02s step size work as well as advertised I would suggest to provide them as a sort of "2nd default". @chraibi whats your opinion on that?

**Reply by chraibi** [reply]
I'm also curious about the suggested changes of the parameters. 
Carole and Jette: Can you please clarify?

Besides, I was wondering, how often do you get the error "agents outside the walkable area" ?

**Reply by Carole1004** [reply]
@Ozaq @chraibi 
Yes, I am happy to clarify that. The error "agents outside the walkable area" happened **really often** at the beginning of my study. 

Mohcine did some trial and error and suggested me these parameters:
strength_neighbour_repulsion = default
range_neightbour_repulsion = default
strength_geometry_repulsion = 26
range_geometry_repulsion = 0.01
It is the reason why they became my **reference** for the study. 

Starting from those, I tried to find the best results while increasing the time step dt. And the result for **dt = 0.02 s** is:
strength_neighbour_repulsion = default
range_neightbour_repulsion = default
strength_geometry_repulsion = 35
range_geometry_repulsion = 0.019

**Reply by Carole1004** [reply]
I hope it is clear now!

**Reply by chraibi** [reply]
Carole that's clear to me.
It would be interesting to know how often the error "The error "agents outside the walkable area" occurs when increasing dt.
